Variable docstring synchronization between command line help and manual

omus opened the pull request:
Uses new Sphinx directive "variable" to synchronize docstrings for variables with the manual. I created several new docstrings based upon the descriptions in the manual.

Requires https://github.com/JuliaLang/JuliaDoc/pull/22 (merged)
Partially addresses #16303.

omus committed:
Improved sync between manual and command line help

Uses new "variable" directive to synchronize variable docstrings to the
manual. Added docstrings for several constants.

omus commented:
Currently `nothing`, `ANY`, `JULIA_HOME`, and `CPU_CORES` are in the helpdb file. I should be able to  get `JULIA_HOME` and `CPU_CORES` out of helpdb. I'm not sure where `nothing` or `ANY` should go as they aren't explicitly defined in base. Suggestions are welcome.

omus commented:
cc @MichaelHatherly: I've modified `genstdlib.jl`

MichaelHatherly commented:
> in the helpdb file

Perhaps in `basedocs.jl`? That's where other docstrings without declarations have been going. Note that instead of `:nothing` etc. you can just put `nothing`. No need to quote the object, it should have the same effect.

omus committed:
Update JuliaDoc revision

tkelman commented on base/version.jl:
```diff
@@ -208,6 +208,13 @@ catch e
+"""
+    VERSION
+
+An object describing which version of Julia is in use.
```
A `VersionNumber` object

omus replied:
Thanks. Good catch.

tkelman replied:
further down it looks like you were just porting over the existing rst description, so consider this an enhancement :)

omus committed:
Move `ANY` and `nothing` docstrings

omus committed:
Correction to VERSION docstring

omus committed:
Add detail reference to VERSION docstring

omus committed:
Move more variable docstrings

omus committed:
Fix broken constant references

omus committed:
Move docstrings for Libc and Libdl constants

omus committed:
Move CPU_CORES docstring

omus committed:
Move DL_LOAD_PATH docstring

omus commented:
I have updated all the docstrings for constants/variables listed in [stdlib/constants](http://julia.readthedocs.io/en/latest/stdlib/constants/). Additionally I noticed most of the "See also" references were broken so I fixed that as well.

I'm sure I have missed some constants that should be updated but I think this PR has modified enough files ;)

MichaelHatherly commented on base/deprecated.jl:
```diff
@@ -612,6 +612,21 @@ end
+        """ ->
+        $constant
+    end
+end
```
Rather than having 3 similar docstrings perhaps something like the following?

```julia
const MS_ASYNC = 1
const MS_INVALIDATE = 2
const MS_SYNC = 4

@doc """
    MS_ASYNC
    MS_INVALIDATE
    MS_SYNC

Enum constants for [`msync`](:func:`msync`). See your platform man page for details.
(not available on Windows).
""" ->
(MS_ASYNC, MS_INVALIDATE, MS_SYNC)
```

and then adjust the signature in the rst docs accordingly.

(`@doc` needed since this is inside a `@unix_only` block, not toplevel.)

omus replied:
How do I get this to match up with the manual?

```
WARNING: duplicate signature found for 'Base.MS_INVALIDATE' in module 'Base':

    MS_ASYNC
    MS_SYNC
    MS_INVALIDATE

WARNING: duplicate signature found for 'Base.MS_ASYNC' in module 'Base':

    MS_ASYNC
    MS_SYNC
    MS_INVALIDATE
```

MichaelHatherly replied:
Adding a

```
.. variable:: MS_ASYNC
              MS_INVALIDATE
              MS_SYNC
```

declaration should be sufficient I think.

omus replied:
Seems so obvious now... I still get the warnings however.

MichaelHatherly replied:
Actually, that's probably a bug in `genstdlib`, since the docstrings are the same object rather than different docstrings that happen to have the same signature. I'll try track that one down.

MichaelHatherly replied:
Could you try applying this diff and see if the duplicate warnings are gone?

```diff
diff --git a/doc/genstdlib.jl b/doc/genstdlib.jl
index b03a146..85e28fb 100644
--- a/doc/genstdlib.jl
+++ b/doc/genstdlib.jl
@@ -111,7 +111,7 @@ function loaddocs!(state::State, mod, binding::Binding, typesig, docstr)
     markdown = Base.Docs.parsedoc(docstr)
     if validdocstr(markdown)
         code = rstrip(markdown.content[1].code)
-        if haskey(state.validdocs, code)
+        if haskey(state.validdocs, code) && state.validdocs[code][3] !== docstr
             code = indent(code)
             warn("duplicate signature found for '$binding' in module '$mod':\n\n$code\n")
             state.errorlevel = 2
```

omus replied:
Patch does fix the duplicate signature warnings. Should I include the patch as part of this PR?

MichaelHatherly replied:
Yes, please do. I don't want to hold up your PR with different one line PR :)

MichaelHatherly commented on base/initdefs.jl:
```diff
@@ -53,6 +72,11 @@ function init_bind_addr()
+        JULIA_HOME
+
+    A string containing the full path to the directory containing the `julia` executable.
+    """ ->
```
Would be good to avoid defining docstrings inside functions where possible. Maybe the following would be better?

```julia
"""
    JULIA_HOME

A string containing the full path to the directory containing the `julia` executable.
"""
:JULIA_HOME

function early_init()
    global const JULIA_HOME = ccall(:jl_get_julia_home, Any, ())
    # ...
```

This works since docstrings are associated with `Binding` objects rather than directly with the real object. We only ever check whether it is defined at runtime when retrieving docs.

(Technically you can write `JULIA_HOME` instead of `:JULIA_HOME` and it will also work, but perhaps the quoted form does better at alerting the read of the "lazy binding" taking place.)

MichaelHatherly commented on base/libdl.jl:
```diff
@@ -20,6 +26,19 @@ const RTLD_NOLOAD    = 0x00000020
+        """ ->
+        $constant
+    end
+end
```
Same comment as above about using the multidoc syntax.

MichaelHatherly commented on base/libuv.jl:
```diff
@@ -99,8 +99,26 @@ function reinit_stdio()
+        STDERR
+
+    Global variable referring to the standard error stream.
+    """ ->
```
Same comment as above about writing the docs outside the function.

MichaelHatherly commented on base/sysinfo.jl:
```diff
@@ -23,6 +23,11 @@ global CPU_CORES
+        CPU_CORES
+
+    The number of CPU cores in the system.
+    """ ->
```
Same comment as above about writing the docs outside the function.

MichaelHatherly commented on base/version.jl:
```diff
@@ -208,6 +208,14 @@ catch e
+A `VersionNumber` object describing which version of Julia is in use. For details see
+[Version Number Literals](:ref:`man-version-number-literals`).
+"""
+:VERSION
```
Can use the unquoted form here, though perhaps rewriting the `try` block would work too. Something like:

```julia
"""
    VERSION

A `VersionNumber` object describing which version of Julia is in use. For details see
[Version Number Literals](:ref:`man-version-number-literals`)
"""
const VERSION = try
    build_number = GIT_VERSION_INFO.build_number != 0 ? "+$(GIT_VERSION_INFO.build_number)" : ""
    convert(VersionNumber, "$(VERSION_STRING)$(build_number)")
catch e
    println("while creating Base.VERSION, ignoring error $e")
    VersionNumber(0)
end
```

(Perhaps there's a reason it's written like it is?)

omus replied:
Good idea. I tend to forget about that syntax

omus committed:
Combine duplicate docstrings

omus committed:
Move docstrings outside of functions

omus committed:
Define VERSION docstring at creation
